src/Body.ts

```diff
@@ -1,44 +1,590 @@
+	 */
+	protected estimatedRowCount(): number {
+		const {
+			_scrollerElement: scrollerElement
+		} = this;
+
+		const height = scrollerElement.offsetHeight;
+		if (height) {
+			return Math.round(height / this.estimatedRowHeight());
+		}
+		return 30;
+	}
+
+	/**
+	 * Based on what is currently rendered, find the average height
+	 * of each row and, if nothing is rendered, return 20.
+	 */
+	protected estimatedRowHeight(): number {
+		const {
+			_itemElementMap: itemElementMap
+		} = this;
+
+		let rowHeight = 0;
+		let rowCount = 0;
+		for (let details of from(itemElementMap.values())) {
+			if (details.element) {
+				rowHeight += details.element.offsetHeight;
+				rowCount++;
+			}
+		}
+
+		return Math.round(rowHeight / rowCount) || 20;
+	}
+
+	/**
+	 * Shared method used by both onElementCreated and
+	 * onElementUpdated to track DOM changes
+	 */
+	protected onElementChange(element: HTMLElement, key: string): void {
+		const {
+			_itemElementMap: itemElementMap,
+			_marginTop: marginTop
+		} = this;
+
+		if (key === 'marginTop') {
+			marginTop && (marginTop.element = element);
+			return;
+		}
+
+		if (key === 'marginBottom') {
+			return;
+		}
+
+		if (key === 'scroller') {
+			this._scrollerElement = element;
+
+			const {
+				properties: {
+					items,
+					slice: {
+						start = 0
+					} = {},
+					onSliceRequest
+				}
+			} = this;
+
+			if (items.length === 0) {
+				// If there has been no data passed (e.g. during initialization),
+				// we wait until the scroll area appears to get a more accurate
+				// estimate of how many rows to ask for initially
+
+				// Use the start value we found and request an amount of data
+				// equal to the additional data above the scroll area, the number
+				// of visible rows and the additional data below the scroll area
+				onSliceRequest && onSliceRequest({ start, count: this.estimatedRowCount() });
+			}
+			else {
+				// We hit this when the children of the scroll area change
+				// e.g. after we guess how many rows we'll need. The onScroll
+				// method will take this information and add additional rows
+				// before and after this content as padding.
+				this.onScroll();
+			}
+
+			return;
+		}
+
+		const details = itemElementMap.get(key);
+		if (details) {
+			// store the DOM node if its representation changes
+			details.element = element;
+		}
+	}
+
+	protected onElementCreated(element: HTMLElement, key: string): void {
+		this.onElementChange(element, key);
+	}
+
+	protected onElementUpdated(element: HTMLElement, key: string): void {
+		if (key === 'scroller') {
+			const {
+				_firstVisibleKey: firstVisibleKey,
+				_itemElementMap: itemElementMap,
+				_marginTop: marginTop,
+				_scrollerElement: scrollerElement,
+				properties: {
+					onScrollToComplete,
+					scrollTo
+				}
+			} = this;
+			let scrollTop = this._scrollTop;
+
+			const detailsEntries = from(itemElementMap.entries());
+
+			// Check to see if all items are new
+			let cleared = true;
```

Comment:
I think I'm following the logic here, but could `cleared` benefit from a more descriptive name?